src/knot_thing_main.c

```diff
@@ -6,196 +6,160 @@
  * of the BSD license. See the LICENSE file for details.
  *
  */
-#include <errno.h>
-#include <stdint.h>
 
 #include <stdint.h>
 #include <stdio.h>
 #include <errno.h>
+#include <string.h>
 
 #include "knot_thing_config.h"
 #include "knot_types.h"
 #include "knot_thing_main.h"
 
-static struct sensor_integer *isensor[KORE_INTEGER_SENSORS];
-static struct sensor_float *fsensor[KORE_FLOAT_SENSORS];
-static struct sensor_bool *bsensor[KORE_BOOL_SENSORS];
-static struct sensor_raw *rsensor[KORE_RAW_SENSORS];
-
-static int8_t isensor_num = 0;
-static int8_t fsensor_num = 0;
-static int8_t bsensor_num = 0;
-static int8_t rsensor_num = 0;
-
-/* Hash used to track value changes: stores last value hash */
-static int32_t ihash[KORE_INTEGER_SENSORS];
-static int32_t fhash[KORE_FLOAT_SENSORS];
-static int8_t bhash[KORE_BOOL_SENSORS];
-static int32_t rhash[KORE_RAW_SENSORS];
-
-static inline int32_t hash(int32_t value_int, int32_t multiplier,
-							int32_t value_dec)
+// TODO: normalize all returning error codes
+
+#define KNOT_THING_EMPTY_ITEM		"EMPTY ITEM"
+#define EMPTY_STRING			((uint8_t *)"")
+
+static struct {
+	// schema values
+	uint8_t			value_type;	// KNOT_VALUE_TYPE_* (int, float, bool, raw)
+	uint8_t			unit;		// KNOT_UNIT_*
+	uint16_t		type_id;	// KNOT_TYPE_ID_*
+	const char *		name;		// App defined data item name
```

Comment:
Defining as pointer may be dangerous: suppose that the developer defined a vector inside the scope of a function. By other-hand, copy the value is inefficient due memory footprint.